File: python/discover/fallback/poisbinom.py
import numpy


try:
    range = xrange
except NameError:
    # This is Python 3
    pass


def poisbinom(p, K):
    if K > len(p):
        return 0.0, 1.0
    
    logp = numpy.log(p)
    lognotp = numpy.log(1 - p)

    memory = numpy.empty((2, len(p)))

    #k = 0
    memory[0, 0] = lognotp[0]
    for j in xrange(1, len(p)):
        memory[0, j] = lognotp[j] + memory[0, j - 1]

    prev = 0
    curr = 1

    cdf = memory[0, len(p) - 1]

    for k in xrange(1, K + 1):
        if k > 1:
            memory[curr, k - 1] = logp[k - 1] + memory[prev, k - 2]
        else:
            memory[curr, k - 1] = logp[k - 1]
        
        for j in xrange(k, len(p)):
            memory[curr, j] = numpy.logaddexp(
                lognotp[j] + memory[curr, j - 1],
                logp[j] + memory[prev, j - 1])

        cdf = numpy.logaddexp(cdf, memory[curr, len(p) - 1])

        curr, prev = prev, curr

    return numpy.exp(memory[prev, len(p) - 1]), numpy.exp(cdf)


def cdf(p, x):
    return poisbinom(p, x)[1]

File: python/src/poisbinom.f90
module poisbinom

  use types, only: dp

  implicit none

contains

  function cdf(p, x)
    use logspace, only: logadd, log1p

    real(dp), dimension(:), intent(in) :: p
    integer, intent(in) :: x
    real(dp) :: cdf

    real(dp), dimension(size(p)) :: logp, lognotp
    real(dp), dimension(2, size(p)) :: memory

    integer :: j, k, n, prev, curr, tmp

    n = size(p)

    if (x > n) then
      cdf = 1
    else if (x < 0) then
      cdf = 0
    else
      logp = log(p)
      lognotp = log1p(-p)

      memory(1, 1) = lognotp(1)
      do j = 2, size(p)
        memory(1, j) = lognotp(j) + memory(1, j - 1)
      end do

      prev = 1
      curr = 2

      cdf = memory(1, n)

      do k = 1, x
         if (k > 1) then
            memory(curr, k) = logp(k) + memory(prev, k - 1)
         else
            memory(curr, k) = logp(k)
         end if

         do j = k + 1, n
            memory(curr, j) = logadd(lognotp(j) + memory(curr, j - 1), &
                                     logp(j) + memory(prev, j - 1))
         end do

         cdf = logadd(cdf, memory(curr, n))
         if (cdf >= 0) then
            cdf = 0
            exit
         end if

         tmp = curr
         curr = prev
         prev = tmp
      end do

      cdf = exp(cdf)
    end if
  end function cdf


  function fullCdf(p, x) result (cdf)
    use logspace, only: logadd, log1p

    real(dp), dimension(:), intent(in) :: p
    integer, intent(in) :: x
    real(dp), dimension(0:x) :: cdf

    real(dp), dimension(size(p)) :: logp, lognotp
    real(dp), dimension(2, size(p)) :: memory

    integer :: j, k, n, prev, curr, tmp

    n = size(p)

    if (x > n) then
      cdf = 1.0
    else
      logp = log(p)
      lognotp = log1p(-p)

      memory(1, 1) = lognotp(1)
      do j = 2, size(p)
        memory(1, j) = lognotp(j) + memory(1, j - 1)
      end do

      prev = 1
      curr = 2

      cdf(0) = memory(1, n)

      do k = 1, x
         if (k > 1) then
            memory(curr, k) = logp(k) + memory(prev, k - 1)
         else
            memory(curr, k) = logp(k)
         end if

         do j = k + 1, n
            memory(curr, j) = logadd(lognotp(j) + memory(curr, j - 1), &
                                     logp(j) + memory(prev, j - 1))
         end do

         cdf(k) = logadd(cdf(k - 1), memory(curr, n))
         if (cdf(k) >= 0) then
            cdf(k:) = 0
            exit
         end if

         tmp = curr
         curr = prev
         prev = tmp
      end do

      cdf = exp(cdf)
    end if
  end function fullCdf

end module poisbinom
